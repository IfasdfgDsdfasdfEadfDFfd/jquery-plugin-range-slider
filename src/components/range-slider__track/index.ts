import './styles/index.css';

export * from './views';
export * from './track.model';
export * from './track.controller';
export * from './track.component';
