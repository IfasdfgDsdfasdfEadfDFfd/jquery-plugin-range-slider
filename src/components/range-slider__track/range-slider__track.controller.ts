import { Controller } from '@core';

class TrackController extends Controller {}

export { TrackController };
