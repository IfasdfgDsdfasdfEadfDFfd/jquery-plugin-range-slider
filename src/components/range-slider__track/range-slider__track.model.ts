import { Model } from '@core';

class TrackModel extends Model {
  name = 'track';

  data = {
    bars: [{ leftOffset: '25%', rightOffset: '75%' }],
  };
}

export { TrackModel };
