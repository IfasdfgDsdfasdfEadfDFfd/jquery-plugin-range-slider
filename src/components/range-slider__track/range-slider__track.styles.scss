@import '../../variables.scss';

.range-slider__track {
  position: absolute;

  top: $top-offset;

  width: 100%;
  height: $track-height;

  background-color: #fff;
  border: 1px solid #242426;
  border-radius: 0.25rem;
}
