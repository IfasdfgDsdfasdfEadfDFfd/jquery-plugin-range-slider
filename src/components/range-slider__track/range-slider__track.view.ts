import { View } from '@core';
import { memo } from 'core/utils';

class Track extends View<TrackProps> {
  attrs = {
    class: 'range-slider__track',
  };
  children = {
    item: new TrackProgress(),
  };

  render({ progressProps }: TrackProps): void {
    this.children.item.render(progressProps);
  }
}

@memo(['updateColor', 'updateLeftOffset', 'updateRightOffset'])
class TrackProgress extends View<TrackProgressProps> {
  attrs = {
    class: 'range-slider__track__progress',
  };

  render({ color, leftOffset, rightOffset }: TrackProgressProps): void {
    this.updateLeftOffset(leftOffset);
    this.updateRightOffset(rightOffset);
    this.updateColor(color);
  }

  updateColor(newColor: string): void {
    this.nativeElement.style.setProperty('background-color', newColor);
  }

  updateLeftOffset(value: number): void {
    this.nativeElement.style.setProperty('left', `${value}%`);
  }

  updateRightOffset(value: number): void {
    this.nativeElement.style.setProperty('right', `${value}%`);
  }
}

export { Track, TrackProgress };
