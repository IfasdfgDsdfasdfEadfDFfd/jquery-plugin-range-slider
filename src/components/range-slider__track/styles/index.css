.range-slider__track {
  position: absolute;
  width: 100%;
  height: var(--progress-height);
  margin: var(--track-offset) 0;
  border-radius: var(--track-border-radius);
  background-color: var(--gray-color);
}

.range-slider__progress {
  position: absolute;
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
}

.range-slider__progress__segment {
  position: absolute;
  height: 100%;
}

.range-slider__thumb {
  position: absolute;
  right: 0;
  left: 0;
  display: flex;
  align-items: center;
  width: 100%;
  height: var(--track-height);
  margin-top: calc((var(--track-offset)) * -1);
}

.range-slider__thumb__item {
  position: absolute;
  width: var(--thumb-width);
  height: var(--thumb-height);
  border: var(--thumb-border);
  border-radius: var(--thumb-border-radius);
  background-color: var(--primary-color);
}

.range-slider__thumb__item__marker {
  display: flex;
  justify-content: center;
  width: fit-content;
  margin: auto;
  margin-top: -1.8rem;
  padding: 0.1rem 0.4rem;
  border: var(--thumb-border);
  border-radius: var(--thumb-marker-border-radius);
  color: var(--thumb-marker-color);
  font-weight: 600;
}

.range-slider__scale {
  position: relative;
  display: flex;
  flex-flow: row nowrap;
  margin: 0;
  padding: var(--track-height) 0;
  list-style: none;
}

.range-slider__scale__item {
  position: absolute;
}

.range-slider__scale__button {
  width: 100%;
  padding: 0;
  border: none;
  border-color: var(--gray-color);
  background-color: transparent;
  color: var(--gray-color);
  font-weight: 600;
  cursor: pointer;

  &::after {
    content: "";
    position: absolute;
    bottom: 120%;
    left: calc(50% - 1px);
    height: 70%;
    border-color: inherit;
    border-right: 1px solid;
    border-left: 1px solid;
    border-radius: 0.4rem;
  }

  &:focus,
  &:hover {
    border-color: inherit;
    color: inherit;
    outline: none;
  }
}
