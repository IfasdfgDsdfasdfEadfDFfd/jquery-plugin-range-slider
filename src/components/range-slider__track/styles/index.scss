@import "../../vars";

.range-slider__track {
  width: 100%;

  &-progress {
    background-color: $default-color;
  }

  &-scale {
    &__item {
      color: $gray-color;
    }

    &__item-button {
      color: inherit;
    }
  }
}
