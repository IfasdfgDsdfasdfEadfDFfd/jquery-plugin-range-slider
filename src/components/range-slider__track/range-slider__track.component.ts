import { Component } from '@core';

import { Track } from './range-slider__track.view';
import { TrackModel } from './range-slider__track.model';
import { TrackController } from './range-slider__track.controller';

class TrackComponent extends Component {
  view = new Track();
  model = new TrackModel();
  controller = new TrackController();
}

export { TrackComponent };
