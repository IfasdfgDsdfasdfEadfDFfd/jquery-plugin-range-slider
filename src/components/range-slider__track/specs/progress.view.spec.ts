import { Progress, ProgressSegment } from 'components/range-slider__track';

describe('Progress', () => {
  let progress: Progress;

  const color = 'rgb(133, 133, 133)';
  const leftOffset = 25;
  const rightOffset = 75;

  const props: ProgressProps = {
    color,
    segments: [
      {
        leftOffset,
        rightOffset,
      },
    ],
  };

  beforeEach(() => {
    progress = new Progress();
    progress.init();
    progress.render(props);
  });

  test('to match snapshot', () => {
    expect(progress.nativeElement).toMatchSnapshot();
  });
});

describe('ProgressSegment', () => {
  let progressSegment: ProgressSegment;

  const color = 'rgb(133, 133, 133)';
  const leftOffset = 25;
  const rightOffset = 75;

  const props: ProgressSegmentProps = {
    color,
    leftOffset,
    rightOffset,
  };

  beforeEach(() => {
    progressSegment = new ProgressSegment();
    progressSegment.init();
    progressSegment.render(props);
  });

  test('updateColor()', () => {
    expect(progressSegment.nativeElement.style.getPropertyValue('background-color')).toEqual(color);

    const newColor = 'rgb(255, 255, 255)';
    expect(newColor).not.toEqual(color);

    progressSegment.updateColor(newColor);
    expect(progressSegment.nativeElement.style.getPropertyValue('background-color')).toEqual(
      newColor,
    );
  });

  test('updateLeftOffset()', () => {
    let receivedOffset = parseFloat(progressSegment.nativeElement.style.getPropertyValue('left'));
    expect(receivedOffset).toEqual(leftOffset);

    const newOffset = 30;
    expect(newOffset).not.toEqual(leftOffset);

    progressSegment.updateLeftOffset(newOffset);
    receivedOffset = parseFloat(progressSegment.nativeElement.style.getPropertyValue('left'));
    expect(receivedOffset).toEqual(newOffset);
  });

  test('updateRightOffset()', () => {
    let receivedOffset = parseFloat(progressSegment.nativeElement.style.getPropertyValue('right'));
    expect(receivedOffset).toEqual(100 - rightOffset);

    const newOffset = 80;
    expect(newOffset).not.toEqual(rightOffset);

    progressSegment.updateRightOffset(newOffset);

    receivedOffset = parseFloat(progressSegment.nativeElement.style.getPropertyValue('right'));
    expect(receivedOffset).toEqual(100 - newOffset);
  });

  test('to match snapshot', () => {
    expect(progressSegment.nativeElement).toMatchSnapshot();
  });
});
