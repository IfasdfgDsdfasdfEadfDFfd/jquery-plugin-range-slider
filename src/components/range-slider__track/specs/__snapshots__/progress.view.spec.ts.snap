// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress to match snapshot 1`] = `
<div
  class="range-slider__progress"
>
  <div
    class="range-slider__progress__segment"
    style="left: 25%; right: 25%; background-color: rgb(133, 133, 133);"
  />
</div>
`;

exports[`ProgressSegment to match snapshot 1`] = `
<div
  class="range-slider__progress__segment"
  style="left: 25%; right: 25%; background-color: rgb(133, 133, 133);"
/>
`;
